### ISSUE TITLE
X-order of data elements changes after signal fires

### ISSUE THREAD

**Issue by tomakita** [vega/vega#354]
Here's something interesting that I noticed today: the specification, below, allows you to change cycle through the dataset by clicking on the left and right sides of the graph viewport.  If you click on the right side (this will show data for Group == 1), and then click on the left side (this will go back to the data for Group == 0), the x-order of the elements in the graph will be changed:

(Note that in all of the three examples that follow, the only thing that changes is the `values` array in the `data` schema)

```javascript
{
  "width": 640,
  "height": 400,
  "signals": [
    {
      "name": "groupNumber",
      "init": { "expr": "0" },
      "streams": [
        {
         "type": "mousedown[eventX() > 320]",
         "field": "Group",
         "expr": "if(groupNumber < 1, groupNumber + 1, groupNumber)"
        },
        {
         "type": "mousedown[eventX() < 320]",
         "expr": "if(groupNumber > 0, groupNumber - 1, groupNumber)"
        }
      ]
    }
  ],
  "data": [
    {
      "name": "employees",
      "values":
              [{"Group":0,"Id":9261,"Value":2},
              {"Group":0,"Id":9270,"Value":1},
              {"Group":1,"Id":9315,"Value":1},
              {"Group":1,"Id":9270,"Value":2}],
      "transform": [
        {"type": "filter", "test": "datum.Group == groupNumber"}
      ]
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "ordinal",
      "range": "width",
      "domain": {"data": "employees", "field": "Id"}
    },
    {
      "name": "y",
      "type": "linear",
      "range": "height",
      "domain": {"data": "employees", "field": "Value"}
    }
  ],
  "axes": [
    {"type": "x", "scale": "x"},
    {"type": "y", "scale": "y"}
  ],
  "marks": [
    {
      "type": "rect",
      "from": {"data": "employees"},
      "properties": {
        "enter": {
          "x": {"scale": "x", "field": "Id"},
          "width": {"scale": "x", "band": true, "offset": -1},
          "y": {"scale": "y", "field": "Value"},
          "y2": {"scale": "y", "value": 0}
        },
        "update": {
          "fill": {"value": "steelblue"}
        },
        "hover": {
          "fill": {"value": "yellow"}
        }
      }
    }
  ]
}
```

If the "Id"s across the two groups are unique, then everything functions as expected, and the x-order is consistent when you cycle through the different Group numbers:

```javascript
{
  "width": 640,
  "height": 400,
  "signals": [
    {
      "name": "groupNumber",
      "init": { "expr": "0" },
      "streams": [
        {
         "type": "mousedown[eventX() > 320]",
         "field": "Group",
         "expr": "if(groupNumber < 1, groupNumber + 1, groupNumber)"
        },
        {
         "type": "mousedown[eventX() < 320]",
         "expr": "if(groupNumber > 0, groupNumber - 1, groupNumber)"
        }
      ]
    }
  ],
  "data": [
    {
      "name": "employees",
      "values":
              [{"Group":0,"Id":9261,"Value":2},
              {"Group":0,"Id":9270,"Value":1},
              {"Group":1,"Id":9315,"Value":1},
              {"Group":1,"Id":9271,"Value":2}],
      "transform": [
        {"type": "filter", "test": "datum.Group == groupNumber"}
      ]
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "ordinal",
      "range": "width",
      "domain": {"data": "employees", "field": "Id"}
    },
    {
      "name": "y",
      "type": "linear",
      "range": "height",
      "domain": {"data": "employees", "field": "Value"}
    }
  ],
  "axes": [
    {"type": "x", "scale": "x"},
    {"type": "y", "scale": "y"}
  ],
  "marks": [
    {
      "type": "rect",
      "from": {"data": "employees"},
      "properties": {
        "enter": {
          "x": {"scale": "x", "field": "Id"},
          "width": {"scale": "x", "band": true, "offset": -1},
          "y": {"scale": "y", "field": "Value"},
          "y2": {"scale": "y", "value": 0}
        },
        "update": {
          "fill": {"value": "steelblue"}
        },
        "hover": {
          "fill": {"value": "yellow"}
        }
      }
    }
  ]
}
```

Curiously, if you change the "Id"s in the dataset to the numbers 1 through 3, everything functions as expected, and the x-order is consistent when you cycle through the different Group numbers:

```javascript
{
  "width": 640,
  "height": 400,
  "signals": [
    {
      "name": "groupNumber",
      "init": { "expr": "0" },
      "streams": [
        {
         "type": "mousedown[eventX() > 320]",
         "field": "Group",
         "expr": "if(groupNumber < 1, groupNumber + 1, groupNumber)"
        },
        {
         "type": "mousedown[eventX() < 320]",
         "expr": "if(groupNumber > 0, groupNumber - 1, groupNumber)"
        }
      ]
    }
  ],
  "data": [
    {
      "name": "employees",
      "values":
              [{"Group":0,"Id":1,"Value":2},
              {"Group":0,"Id":2,"Value":1},
              {"Group":1,"Id":3,"Value":1},
              {"Group":1,"Id":2,"Value":2}],
      "transform": [
        {"type": "filter", "test": "datum.Group == groupNumber"}
      ]
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "ordinal",
      "range": "width",
      "domain": {"data": "employees", "field": "Id"}
    },
    {
      "name": "y",
      "type": "linear",
      "range": "height",
      "domain": {"data": "employees", "field": "Value"}
    }
  ],
  "axes": [
    {"type": "x", "scale": "x"},
    {"type": "y", "scale": "y"}
  ],
  "marks": [
    {
      "type": "rect",
      "from": {"data": "employees"},
      "properties": {
        "enter": {
          "x": {"scale": "x", "field": "Id"},
          "width": {"scale": "x", "band": true, "offset": -1},
          "y": {"scale": "y", "field": "Value"},
          "y2": {"scale": "y", "value": 0}
        },
        "update": {
          "fill": {"value": "steelblue"}
        },
        "hover": {
          "fill": {"value": "yellow"}
        }
      }
    }
  ]
}
```

Any idea what might be causing this one?  Thanks for Vega -- it's awesome!!

**Comment by arvind** [comment]
Hi @tomakita, the sort order of ordinal scales is not guaranteed by default. Instead, you can specify an [explicit order](https://github.com/vega/vega/wiki/Scales#sorting-domains-for-ordinal-and-quantile-scales) within the domain's data reference. (Note, setting sort to `true` is currently available in commit 1150aba, and will be available in the next release of Vega).

**Comment by tomakita** [comment]
Cool, thanks for the info!
